src/sknnr/utils/__init__.py

```diff
@@ -0,0 +1,137 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import TYPE_CHECKING, Any
+
+import numpy as np
+
+if TYPE_CHECKING:
+    import pandas as pd
+
+
+def is_dataframe_like(obj: Any) -> bool:
+    """
+    Check if `obj` is a dataframe-like structure.  This will evaluate to `True`
+    for pandas and polars DataFrames without the need to import those packages
+    explicitly.
+    """
+    return hasattr(obj, "columns") and hasattr(obj, "dtypes")
+
+
+def is_series_like(obj: Any) -> bool:
+    """
+    Check if `obj` is a series-like structure.  This will evaluate to `True`
+    for pandas and polars Series without the need to import those packages
+    explicitly.
+    """
+    return hasattr(obj, "name") and hasattr(obj, "dtype")
+
+
+def is_number_like_type(t: Any) -> bool:
+    """
+    Check if `t` is a number-like type.  For most types, np.issubdtype will
+    correctly identify the type.  For pandas extension types, we can check the
+    kind of the type.
+    """
+    try:
+        return np.issubdtype(t, np.number)
+    except TypeError:
+        try:
+            return t.kind in "iuf"
+        except AttributeError as err:
+            msg = f"Unsupported type {t}"
+            raise TypeError(msg) from err
+
+
+def is_nan_like(x: Any) -> bool:
+    """Check if `x` is NaN-like."""
+    return bool(
+        x is None
+        or (isinstance(x, float) and np.isnan(x))
+        or x.__class__.__name__ == "NAType"
```

Comment:
Worth a quick note that this covers `pd.NA`?

Reply:
I've expanded the docstring to be more explicit here.

Good question on the `float` check.  Strings definitely don't work with `np.isnan` and I guess I was thinking that we're trying to catch `np.nan` (or anything that evaluates to `np.nan - e.g. `np.log(-1)`) and those would always be float type.  Is my thinking correct on that?

Reply:
Makes sense, thanks for explaining!